{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Stitch1_replicate1-1': 2144, 'Stitch1_replicate1-2': 1077, 'Stitch1_replicate2-1': 1485, 'Stitch1_replicate2-2': 2320, 'Stitch2_replicate1-1': 418, 'Stitch2_replicate1-2': 1604, 'Stitch2_replicate2-1': 3015, 'Stitch2_replicate2-2': 896, 'Stitch3_replicate1-1': 84716, 'Stitch3_replicate1-2': 76016, 'Stitch3_replicate2-1': 55020, 'Stitch3_replicate2-2': 43223, 'Stitch4_replicate1-1': 299149, 'Stitch4_replicate1-2': 253614, 'Stitch4_replicate2-1': 684302, 'Stitch4_replicate2-2': 132052}\n"
     ]
    }
   ],
   "source": [
    "#Use this code after running the minimap2 using the code pooled_GPCR_run_minimap2.sh\n",
    "#check how many input reads in the beginning\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "read_summaryD={}\n",
    "sampleL=[]\n",
    "\n",
    "stitch_orderL=[\"Stitch1_replicate\", \"Stitch2_replicate\", \"Stitch3_replicate\", \"Stitch4_replicate\"]\n",
    "repeatL=[\"1-1\",\"1-2\",\"2-1\",\"2-2\"]\n",
    "#1st stitch\n",
    "\n",
    "for i in range (0,4):\n",
    "    for repeat in repeatL:\n",
    "        sampleL.append(stitch_orderL[i]+repeat)\n",
    "        file_path=stitch_orderL[i]+repeat+\".fastq\"\n",
    "        count = sum(1 for record in SeqIO.parse(file_path, \"fastq\"))\n",
    "        read_summaryD[stitch_orderL[i]+repeat]=count\n",
    "        \n",
    "print(read_summaryD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check how many reads mapped to the single ref.\n",
    "#use this to generate Table S8. Read coverage and number of GPCR designs detected at various steps of filtering for the pooled GPCR library assembly data.\n",
    "\n",
    "#run mimimap2 to get this paf file\n",
    "#alalysis based on the alignment results and also compare to the known posistion GPCR\n",
    "\n",
    "import os, re\n",
    "\n",
    "alignD={}\n",
    "assignD={}\n",
    "mappedD={}\n",
    "removeD={}\n",
    "#path = '/Users/po-hsianghung/Documents/GPCR/pool_assay_analysis/raw_demuxed_file/merged_reads/'\n",
    "\n",
    "out1=open(\"reads_mapped_summary.xls\",'w')\n",
    "out2=open(\"GPCR_assemble_summary.xls\",'w')\n",
    "out1.write(\"barcode\"+\"\\t\"+\"total_reads\"+\"\\t\"+\"total_mapped_reads\"+\"\\t\"+\"total_unique_mapped_reads\"+\"\\n\")\n",
    "\n",
    "#for test purpose, using few samples\n",
    "#sampleL=['A1-1','B1-1','C1-1','D1-1']\n",
    "\n",
    "for n_barcode in sampleL:\n",
    "    alignD[n_barcode]={}\n",
    "    mappedD[n_barcode]={}\n",
    "    removeD[n_barcode]={}\n",
    "    n_total=0 #this is mappable reads. It is not equal to the original reads\n",
    "    removeD[n_barcode]=[]\n",
    "    file2=open(\"paf/\"+n_barcode+\".paf\")\n",
    "    #record the best assign into alignD, if there are same socre with multiple hits - go to removeD\n",
    "    for line in file2.readlines():\n",
    "        line=line.strip().split(\"\\t\")\n",
    "        read_name=line[0]\n",
    "        read_length=line[1]\n",
    "        ref_name=line[5]\n",
    "        ref_length=line[6]\n",
    "        n_mapped=int(line[9]) #number of residule mapped\n",
    "        if read_name in alignD[n_barcode].keys():\n",
    "            if alignD[n_barcode][read_name][2] < n_mapped:\n",
    "                alignD[n_barcode][read_name]=[read_length,ref_name, n_mapped]\n",
    "            elif alignD[n_barcode][read_name][2] == n_mapped: #the same, can not distinguish\n",
    "                if read_name not in removeD[n_barcode]:\n",
    "                    removeD[n_barcode].append(read_name)\n",
    "        else:\n",
    "            n_total=n_total+1\n",
    "            alignD[n_barcode][read_name]=[read_length,ref_name, n_mapped]\n",
    "    #count unique mapped reads into reference [mappedD]\n",
    "    for read_name in list(alignD[n_barcode].keys()):\n",
    "        if read_name not in removeD[n_barcode]: \n",
    "            ref_name=alignD[n_barcode][read_name][1]\n",
    "            if ref_name in mappedD[n_barcode].keys():\n",
    "                mappedD[n_barcode][ref_name]=mappedD[n_barcode][ref_name]+1\n",
    "            else:\n",
    "                mappedD[n_barcode][ref_name]=1\n",
    "     \n",
    "    out1.write(str(n_barcode)+\"\\t\"+str(read_summaryD[n_barcode])+\"\\t\"+str(n_total)+\"\\t\"+str(n_total-len(removeD[n_barcode]))+\"\\n\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "gpcr=[\"A\", \"B\", \"C\", \"D\",\"E\",\"F\"]\n",
    "#stitch_orderL=[\"A\", \"B\", \"C\", \"D\"]\n",
    "gpcrL=[]\n",
    "\n",
    "for a in gpcr:\n",
    "   gpcrL.append(a)\n",
    "\n",
    "for a in gpcr:\n",
    "    for b in gpcr:\n",
    "        gpcrL.append(a+b)\n",
    "\n",
    "for a in gpcr:\n",
    "    for b in gpcr:\n",
    "        for c in gpcr:\n",
    "            gpcrL.append(a+b+c)\n",
    "\n",
    "for a in gpcr:\n",
    "    for b in gpcr:\n",
    "        for c in gpcr:            \n",
    "            for d in gpcr:\n",
    "                gpcrL.append(a+b+c+d)\n",
    "\n",
    "\n",
    "\n",
    "#output the out2 file - GPCR reads information\n",
    "\n",
    "\n",
    "out2.write(\"GPCR\"+\"\\t\"+\"stitch\"+\"\\t\")\n",
    "for sample in sampleL:\n",
    "    out2.write(sample+\"\\t\")\n",
    "out2.write(\"\\n\")\n",
    "\n",
    "resultD={}\n",
    "for ref in gpcrL:\n",
    "    resultD[ref]=['.']*len(sampleL)\n",
    "    for sample in sampleL:\n",
    "        \n",
    "        if ref in mappedD[sample].keys():\n",
    "            resultD[ref][sampleL.index(sample)]=str(mappedD[sample][ref])\n",
    "        else:\n",
    "            resultD[ref][sampleL.index(sample)]=\"0\"\n",
    "\n",
    "for ref in gpcrL:\n",
    "    out2.write(ref+\"\\t\"+str(len(ref))+\"\\t\")\n",
    "    for num in resultD[ref]:\n",
    "        out2.write(num+\"\\t\")\n",
    "    out2.write(\"\\n\")\n",
    "\n",
    "\n",
    "out2.close()\n",
    "out1.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for plot\n",
    "#first, add expected values and parnet information for making plots\n",
    "#use this to generate Table S10. Data analysis summary for the pooled GPCR assembly. \n",
    "\n",
    "file2=open(\"GPCR_assemble_summary.xls\")\n",
    "\n",
    "out1=open(\"pool_summary_forplot.csv\",'w')\n",
    "out1.write(\"unique id\"+\",\"+\"GPCR\"+\",\"+\"parent\"+\",\"+\"#stitch\"+\",\"+\"1st\"+\",\"+\"2nd\"+\",\"+\"3rd\"+\",\"+\"4th\"+\",\"+\"R_1-1\"+\",\"+\"R_1-2\"+\",\"+\"R_2-1\"+\",\"+\"R_2-2\"+\n",
    "\",frq_1-1,frq_1-2,frq_2-1,frq_2-2,exp_1-1,exp_frq_1-1,exp_1-2,exp_frq_1-2,exp_2-1,exp_frq_2-1,exp_2-2\" +   \",exp_frq_2-2,\"+\"\\n\")\n",
    "\n",
    "\n",
    "#for lineage plot\n",
    "#add expected values and parnet information for making plots\n",
    "out2=open(\"pool_summary_forplot_lineage.csv\",'w')\n",
    "out2.write(\"unique id\"+\",\"+\"GPCR\"+\",\"+\"parent\"+\",\"+\"#stitch\"+\",\"+\"1st\"+\",\"+\"2nd\"+\",\"+\"3rd\"+\",\"+\"4th\"+\",\"+\"R_1-1\"+\",\"+\"R_1-2\"+\",\"+\"R_2-1\"+\",\"+\"R_2-2\"+\n",
    "\",frq_1-1,frq_1-2,frq_2-1,frq_2-2,exp_1-1,exp_frq_1-1,exp_1-2,exp_frq_1-2,exp_2-1,exp_frq_2-1,exp_2-2\" +   \",exp_frq_2-2,\"+\"\\n\")\n",
    "\n",
    "\n",
    "#input the 0 stitch - recipeint only for lineage tracing.\n",
    "out2.write(\"-1\"+\",\"+\"R\"+\",\"+\"NA\"+\",\"+\"0\"+\",\"+\"NA\"+\",\"+\"NA\"+\",\"+\"NA\"+\",\"+\"NA\"+\",\"+\"1\"+\",\"+\"1\"+\",\"+\"1\"+\",\"+\"1\"+\n",
    "\",1,1,1,1,1,1,1,1,1,1,1,1,\"+\"\\n\")\n",
    "\n",
    "\n",
    "resultD={}\n",
    "\n",
    "#GPCRL=gpcrL\n",
    "\n",
    "for line in file2.readlines():\n",
    "    if line.startswith(\"GPCR\"):\n",
    "        line=line.strip().split(\"\\t\")\n",
    "        head=line\n",
    "    else:\n",
    "        line=line.strip().split(\"\\t\")\n",
    "        #resultD[gpcrL.index(line[0])]=[line[0], gpcrL.index(line[0][0:3]) ,len(line[0])] + list(line[0])+[\"NA\"]*(4-len(line[0]))+line[1:5]\n",
    "        if len(line[0])==1: #1st stitch\n",
    "            resultD[gpcrL.index(line[0])]=[line[0],-1,len(line[0])] + list(line[0])+[\"NA\"]*(4-len(line[0]))+line[2:6]\n",
    "        elif len(line[0])==2: #2nd stitch\n",
    "            resultD[gpcrL.index(line[0])]=[line[0],  gpcrL.index(line[0][0]) ,len(line[0])] + list(line[0])+[\"NA\"]*(4-len(line[0]))+line[6:10]\n",
    "        elif len(line[0])==3: #3nd stitch\n",
    "            resultD[gpcrL.index(line[0])]=[line[0], gpcrL.index(line[0][0:2])   ,len(line[0])] + list(line[0])+[\"NA\"]*(4-len(line[0]))+line[10:14]\n",
    "        elif len(line[0])==4: #4th stitch\n",
    "            resultD[gpcrL.index(line[0])]=[line[0], gpcrL.index(line[0][0:3]) ,len(line[0])] + list(line[0])+[\"NA\"]*(4-len(line[0]))+line[14:18]\n",
    "        else:\n",
    "            print (\"error: only 4 stitch, \", line)\n",
    "\n",
    "\n",
    "\n",
    "totalD={}\n",
    "for i in range (0, len(resultD)):\n",
    "    stitch=resultD[i][2]\n",
    "    if stitch in totalD.keys():\n",
    "        for j in range (0,4):\n",
    "            totalD[stitch][j]=int(totalD[stitch][j])+int(resultD[i][7:11][j])\n",
    "    else:\n",
    "        totalD[stitch]=resultD[i][7:11]\n",
    "\n",
    "#calculate the frequency for each stitch\n",
    "for i in range (0, len(resultD)):\n",
    "    stitch=resultD[i][2]\n",
    "    n_total=totalD[stitch]\n",
    "    for j in range (0, len(resultD[i][7:11])):\n",
    "        #print (resultD[i][7+j] , n_total[j])\n",
    "        resultD[i].append(     int(resultD[i][7+j])     /n_total[j] )\n",
    "#calculate the expected value\n",
    "for n in range (0, len(resultD)):\n",
    "    stitch=resultD[n][2]\n",
    "    n_total=totalD[stitch]\n",
    "    parent=resultD[n][1]\n",
    "    exp_frq=float(1)/6\n",
    "    \n",
    "    if parent==-1:\n",
    "        p_frq=[1,1,1,1]\n",
    "    else:\n",
    "        p_frq=resultD[parent][11:15] #parental real frq.\n",
    "\n",
    "    for j in range (0, len(p_frq)):\n",
    "        resultD[n].append(p_frq[j] * exp_frq * n_total[j])\n",
    "        resultD[n].append(p_frq[j] * exp_frq)\n",
    "\n",
    "for i in range (0, len(resultD)):\n",
    "    out1.write(str(i)+\",\")\n",
    "    out2.write(str(i)+\",\")\n",
    "    for item in resultD[i]:\n",
    "        out1.write(str(item)+\",\")\n",
    "        out2.write(str(item)+\",\")\n",
    "    out1.write(\"\\n\")\n",
    "    out2.write(\"\\n\")\n",
    "\n",
    "out1.close()    \n",
    "out2.close()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify missing GPCRs, whether the missing is due to the parental GPCR is failed assembled in previous round\n",
    "# use this to generate Table S9. Table of missed GPCR variants. \n",
    "file5=open(\"pool_summary_forplot_lineage.csv\")\n",
    "\n",
    "out1=open(\"missing.xls\",'w')\n",
    "\n",
    "\n",
    "idD={}\n",
    "for line in file5.readlines():\n",
    "    if not line.startswith(\"unique\"):\n",
    "        line=line.strip().split(\",\")\n",
    "        idD[line[0]]=line[1]\n",
    "\n",
    "\n",
    "\n",
    "missD={}\n",
    "\n",
    "repeatL=[\"1-1\",\"1-2\",\"2-1\",\"2-2\"]\n",
    "\n",
    "countD={}\n",
    "\n",
    "\n",
    "file5=open(\"pool_summary_forplot_lineage.csv\")\n",
    "for line in file5.readlines():\n",
    "    if not line.startswith(\"unique\"):\n",
    "        line=line.strip().split(\",\")\n",
    "        n_stitch=int(line[3])\n",
    "        n_countL=line[8:12]\n",
    "        if n_stitch!=0:\n",
    "            if n_stitch in countD.keys():\n",
    "                for i in range (0, 4):\n",
    "                    if int(n_countL[i])!=0:\n",
    "                        countD[n_stitch][i]=countD[n_stitch][i]+1\n",
    "                    else:\n",
    "                        #print (line[1],n_stitch,repeatL[i])\n",
    "                        #out1.write(line[0]+\"\\t\"+line[2]+\"\\t\"+line[1]+\"\\t\"+str(n_stitch)+\"\\t\"+repeatL[i]+\"\\n\")\n",
    "                        if repeatL[i] in missD.keys():\n",
    "                            if n_stitch in missD[repeatL[i]].keys():\n",
    "                                if idD[line[2]] in missD[repeatL[i]][n_stitch].keys():\n",
    "                                    missD[repeatL[i]][n_stitch][idD[line[2]]].append(line[1])\n",
    "                                else:\n",
    "                                    missD[repeatL[i]][n_stitch][idD[line[2]]]=[line[1]]\n",
    "\n",
    "                            else:\n",
    "                                missD[repeatL[i]][n_stitch]={}\n",
    "                                missD[repeatL[i]][n_stitch][idD[line[2]]]=[line[1]]                                \n",
    "                        else:\n",
    "                            missD[repeatL[i]]={}\n",
    "                            missD[repeatL[i]][n_stitch]={}\n",
    "                            missD[repeatL[i]][n_stitch][idD[line[2]]]=[line[1]]\n",
    "                            #print ('1')\n",
    "\n",
    "            else:\n",
    "                countD[n_stitch]=[0,0,0,0]\n",
    "                for i in range (0, 4):\n",
    "                    if int(n_countL[i])!=0:\n",
    "                        countD[n_stitch][i]=countD[n_stitch][i]+1\n",
    "\n",
    "\n",
    "#print (countD)\n",
    "#print (missD)\n",
    "\n",
    "\n",
    "out1.write(\"repeat\"+\"\\t\"+\"stitch\"+\"\\t\"+\"parent\"+\"\\t\"+\"# of missed child\"+\"\\t\"+\"missed child\"+\"\\n\")\n",
    "for repeat in repeatL:\n",
    "    #for stitch in sorted(list(missD[repeat].keys())):\n",
    "    for stitch in list(missD[repeat].keys()):\n",
    "        for parent in missD[repeat][stitch].keys():\n",
    "            out1.write(repeat+\"\\t\"+str(stitch)+\"\\t\"+parent+\"\\t\"+str(len(missD[repeat][stitch][parent]))+\"\\t\")\n",
    "            for item in missD[repeat][stitch][parent]:\n",
    "                out1.write(item+\"\\t\")\n",
    "            out1.write(\"\\n\")      \n",
    "\n",
    "out1.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify unqiue hits\n",
    "file5=open(\"pool_summary_forplot_lineage.csv\")\n",
    "uniqueD={}\n",
    "for line in file5.readlines():\n",
    "    if not line.startswith(\"unique\"):\n",
    "        line=line.strip().split(\",\")\n",
    "        n_stitch=int(line[3])\n",
    "        n_countL=line[8:12]\n",
    "        variants=line[1]\n",
    "        if n_stitch!=0:\n",
    "            if n_stitch in uniqueD.keys():\n",
    "                for i in range (0, 4):\n",
    "                    if int(n_countL[i])!=0:\n",
    "                        if variants not in uniqueD[n_stitch]:\n",
    "                            uniqueD[n_stitch].append(variants)\n",
    "                    \n",
    "            else:\n",
    "                uniqueD[n_stitch]=[]\n",
    "                for i in range (0, 4):\n",
    "                    if int(n_countL[i])!=0:\n",
    "                        if variants not in uniqueD[n_stitch]:\n",
    "                            uniqueD[n_stitch].append(variants)\n",
    "                    \n",
    "\n",
    "for i in range (1,5):\n",
    "    print (i, len(uniqueD[i]))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
